### PR TITLE
Add wrappers for Android logcat broadcaster

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
## Change list

Added basic classes to support web sockets interaction in general and to interact with logcat messages broadcaster that has been implemented for Android server.
 
## Types of changes

- [ ] No changes in production code.
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

**Comment by mykola-mokhnach** [comment]
@KazuCocoa FYI

File: src/main/java/io/appium/java_client/ws/CanHandleMessages.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+ * This interface might be assigned to classes, which
+ * are defined as web socket clients.
+ */
+public interface CanHandleMessages<T extends MessagesHandler> {
+    /**
+     * @return The list of web socket message handlers.
+     */
+    List<T> messageHandlers();
```
Comment (vmaxim) [inline src/main/java/io/appium/java_client/ws/CanHandleMessages.java]: ```messageHandlers()``` -> ```getMessageHandlers()```

File: src/main/java/io/appium/java_client/ws/MessagesHandler.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
+package io.appium.java_client.ws;
+
+/**
+ * This is the basic interface for all web socket message handlers.
+ */
+public interface MessagesHandler {
```
Comment (vmaxim) [inline src/main/java/io/appium/java_client/ws/MessagesHandler.java]: It's strange that messages handler can not handle messages.

File: src/main/java/io/appium/java_client/ws/StringWebSocketClient.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+    /**
+     * This event if fired when the client is
+     * disconnected from a web socket.
+     *
+     * @param session the actual web socket session instance
+     * @param reason connection close reason
+     */
+    @OnClose
+    public void onClose(Session session, CloseReason reason) {
```
Comment (vmaxim) [inline src/main/java/io/appium/java_client/ws/StringWebSocketClient.java]: ```session``` parameter is never used.

Comment (mykola-mokhnach) [reply src/main/java/io/appium/java_client/ws/StringWebSocketClient.java]: it is used in `connect`

File: src/main/java/io/appium/java_client/ws/MessagesHandler.java (hunk 2 above)
Comment (mykola-mokhnach) [reply src/main/java/io/appium/java_client/ws/MessagesHandler.java]: yep, I'll make it generic

**Comment by mykola-mokhnach** [comment]
I was also thinking about having separate callbacks for each event instead of interface implementation, which is more Java 8'ish. for example:

Instead of 

```java
          driver.addLogcatListener(new MessagesHandler<String>() {
                @Override
                public void onMessage(String message) {
                    messageSemaphore.release();
                }

                @Override
                public void onConnected() {
                    connectedSemaphore.release();
                }

                @Override
                public void onDisconnected() {
                    // ignore
                }

                @Override
                public void onError(Throwable cause) {
                    // ignore
                }
            });
```

it might look like

```java
driver.addLogcatMessageListener(this::onMessage);
driver.addLogcatErrorListener(this::onError);
```

File: src/main/java/io/appium/java_client/ws/StringWebSocketClient.java (hunk 3 above)
Comment (vmaxim) [reply src/main/java/io/appium/java_client/ws/StringWebSocketClient.java]:
```
@OnClose
public void onClose(CloseReason reason) {
```

**Comment by mykola-mokhnach** [comment]
Make sure the hotfix https://github.com/appium/appium/pull/10496 is merged to appium server before testing this PR locally

**Comment by SrinivasanTarget** [comment]
@mykola-mokhnach Unfortunately there is a conflict. Can you please resolve this? I will get this in.

**Comment by mykola-mokhnach** [comment]
@SrinivasanTarget The conflict is resolved

**Pull request merged**